### ISSUE TITLE
is there a method to add (write) to the TileDB on-disk (e.g. on-disk rbind/cbind)?

### ISSUE THREAD

**Issue by PedroMilanezAlmeida** [LTLA/TileDBArray#26]
hope the title is self-descriptive.

```
set.seed(123) # fix RNG seed
mat <- matrix(0, nrow=8, ncol=20) # create matrix with zeros
mat[sample(seq_len(8*20), 15)] <- seq(1, 15) # add non-zero entries
spmat <- as(mat, "dgCMatrix") # make it sparse
path <- tempfile() # get temp file path
out <- TileDBArray::writeTileDBArray(x = spmat, # write to location
                                     path = path)
# error bc file path already exist (and prob inefficient [guessing out needs be from disk read?])
out <- TileDBArray::writeTileDBArray(x = cbind(out,out),
                                     path = path)
```
EDIT: example from https://dirk.eddelbuettel.com/papers/useR2021_tiledb_tutorial.pdf

**Comment by eddelbuettel** [comment]
(In TileDB proper it depends on whether the array is dense or sparse.  For dense, we sometimes 'on purpose' set very wide domain ranges to be able to to extend.  So we should be able to accomodate.  Am between three other open issues but happy to chat.   And three cheers for the tutorial by @aaronwolen and myself coming in handy.)

(So can you add `print(tiledb::schema(path))` and show us the defails?)

**Comment by PedroMilanezAlmeida** [comment]
Thanks for the quick reply!

```
tiledb_array_schema(
    domain=tiledb_domain(c(
        tiledb_dim(name="d1", domain=c(1L,8L), tile=8L, type="INT32", filters=tiledb_filter_list(c(tiledb_filter_set_option(tiledb_filter("ZSTD"),"COMPRESSION_LEVEL",-1)))),
        tiledb_dim(name="d2", domain=c(1L,20L), tile=20L, type="INT32", filters=tiledb_filter_list(c(tiledb_filter_set_option(tiledb_filter("ZSTD"),"COMPRESSION_LEVEL",-1))))
    )),
    attrs=c(
        tiledb_attr(name="x", type="FLOAT64", ncells=1, nullable=FALSE)
    ),
    cell_order="COL_MAJOR", tile_order="COL_MAJOR", capacity=10000, sparse=TRUE, allows_dups=FALSE,
    coords_filter_list=tiledb_filter_list(c(tiledb_filter_set_option(tiledb_filter("ZSTD"),"COMPRESSION_LEVEL",-1))),
    offsets_filter_list=tiledb_filter_list(c(tiledb_filter_set_option(tiledb_filter("ZSTD"),"COMPRESSION_LEVEL",-1))),
    validity_filter_list=tiledb_filter_list(c(tiledb_filter_set_option(tiledb_filter("RLE"),"COMPRESSION_LEVEL",-1)))
)
```

**Comment by eddelbuettel** [comment]
The domain inherits the 8 x 20 size from your incoming matrix which is the binding constraint here.  

I have the feeling we have tricks for this in _e.g._ [TileDB-SOMA](https://github.com/single-cell-data/TileDB-SOMA/) now too (but I am blanking a little what it is).    Here `writeTileDBArray()` reads the `dim()` off the incoming object so we would need some generalization (or for you to pre-process the object to make it larger first).  Hmpf.

**Comment by LTLA** [comment]
**tl;dr** I don't have time myself to add something like this but I wouldn't mind entertaining a PR.

Technically speaking, almost all operations in the **DelayedArray** framework are read-only with respect to any file-backed resources. This is necessary to avoid total chaos when, e.g., multiple `TileDBArray` objects refer to the same TileDB directory and changes to the latter result in silent mutations-by-reference to all of the former. So, any modification to the directory must be done explicitly - at which point it is assumed that the user understands the implications - not as part of a function call like `cbind`/`rbind` that would be expected to have no side effects.

In short, appending a `TileDBArray` to another on disk would be its own dedicated function, maybe something like `appendTileDBArrays`. Though one would have to question whether **TileDBArray** really needs to be involved here, if it can be done with the lower-level **tiledb** R package; it's not like we're using any of the **DelayedArray** bindings around TileDB, which is the primary value-add of **TileDBArray** (in allowing TileDB to play nicely with other Bioconductor data structures). Nonetheless, if it is more convenient to do this operation on `TileDBArray` objects, then I suppose it couldn't hurt to have a function like this... so a PR would be welcome.

**Comment by PedroMilanezAlmeida** [comment]
I agree with @LTLA that an explicit method would be preferable rather than just rbind/cbind. Also, I am happy to move this issue to `tiledb`.

@eddelbuettel in case I know the size (dims) of the target file on-disk, can I create an empty (sparse) file of the right size (dims) with `tiledb` and then add data to the respective indices?

I am looking at the following python code chunks and want to try to re-write them in R with `tiledb`:

code source: https://forum.tiledb.com/t/multiple-concurrent-writers-to-append-only-sparse-array/591/2

 ```
#%%
import tiledb
import numpy as np
import tempfile

#%%
uri = tempfile.mkdtemp()

# create a sparse array with dimensions (sample, gene) and attribute (count)
dims = [
  tiledb.Dim(name="sample", domain=(0, 100), tile=10, dtype=np.int32),
  tiledb.Dim(name="gene", domain=(0, 2), tile=None, dtype='ascii')
]
dom = tiledb.Domain(*dims)
attr = tiledb.Attr(name="count", dtype=np.int32)

schema = tiledb.ArraySchema(domain=dom, sparse=True, attrs=[attr])

# create a new array
tiledb.Array.create(uri, schema)
#%%
# write first set of test data
s1 = np.arange(10)
g1 = np.array(['gene1', 'gene2', 'gene3', 'gene4', 'gene5', 'gene6', 'gene7', 'gene8', 'gene9', 'gene10'])
c1 = np.random.randint(0, 100, 10)

with tiledb.open(uri, mode='w') as A:
    A[s1, g1] = c1

# %%
# write second set of test data
s1 = np.arange(10, 20)
g2 = np.array(['gene1', 'gene2', 'gene3', 'gene4', 'gene5', 'gene6', 'gene7', 'gene8', 'gene9', 'gene10'])
c2 = np.random.randint(0, 100, 10)

with tiledb.open(uri, mode='w') as A:
    A[s1, g1] = c1
# %%
with tiledb.open(uri) as A:
    print(A.multi_index[:])

# %%
# slice only the first 5 samples
with tiledb.open(uri) as A:
    print(A.multi_index[:5])

# %%
# slice only the 9th, 13th, and 15-17th samples
with tiledb.open(uri) as A:
    print(A.df[[9, 13, 15, 16, 17]])
# %%
```
Multiprocessor:

```
#%%
import tiledb
import numpy as np
import tempfile
import multiprocessing as mp
if mp.get_start_method(allow_none=True) != 'fork':
  pass #mp.set_start_method('fork')


#%%
def create_array(uri):
    # create a sparse array with dimensions (sample, gene) and attribute (count)
    dims = [
      tiledb.Dim(name="sample", domain=(0, 100), tile=10, dtype=np.int32),
      tiledb.Dim(name="gene", domain=(0, 2), tile=None, dtype='ascii')
    ]
    dom = tiledb.Domain(*dims)
    attr = tiledb.Attr(name="count", dtype=np.int32)

    schema = tiledb.ArraySchema(domain=dom, sparse=True, attrs=[attr])

    # create a new array
    tiledb.Array.create(uri, schema)

#%%
def write_data(uri, samples, genes, counts):
    with tiledb.open(uri, mode='w') as A:
        A[samples, genes] = counts


#%%
def create_data(partitions, partition_size):
    samples = []
    genes = []
    counts = []
    for i,p in enumerate(range(partitions)):
        samples.append(np.arange(i * partition_size, (i+1)*partition_size))
        genes.append(["gene" + str(i) for i in np.random.randint(0, 1000, partition_size)])
        counts.append(np.random.randint(0, 100, partition_size))
    return samples, genes, counts

#%%
def run_batched():
  uri = tempfile.mkdtemp()
  create_array(uri)

  npartitions = 5
  partition_size = 7
  dataset = create_data(npartitions, partition_size)

  # create tuples of data to write on each worker
  items = list(zip([uri for _ in range(npartitions)], *dataset))

  with mp.Pool(4) as pool:
      pool.starmap(write_data, items)

  return uri
# %%
if __name__ == "__main__":
    uri = run_batched()

    with tiledb.open(uri) as A:
      print(A.df[:])
```

**Comment by eddelbuettel** [comment]
@PedroMilanezAlmeida We _definitely_ support appending, and have for some time. My go-to tool is the `tiledb::fromDataFrame()` helper which has an explicit `mode=` argument too.  As you note, you need to ensure the dimensions are sufficient to hold the subsequent writes, and that the array is sparse.  (An example are for example time series indexed by seconds since the epoch:  we have written schemas with dims from 0 to MAX (for the type).  Ditto for geospatial with longitude and latude which have natural lower and upper bounds.)  So yes this will work from any API including R as it is supported in Core.

**Comment by PedroMilanezAlmeida** [comment]
@eddelbuettel 

sorry to continue to use this issue after closing, but, just fyi, when I run the following code, I get into errors with wrong format for `fromDataFrame` (data is a sparse matrix, not a data frame) and unused argument for `fromSparseMatrix` (which does not accept `mode = "append"`). I will prob use `fromDataFrame` (or something else) but now passing the result of `Matrix::summary(object = spmat)` to append `i`, `j` and `x` as a new layer of a 3D cube (assuming that is possible).

```
library(tiledb)
library(Matrix)

set.seed(123) # fix RNG seed
mat <- matrix(data = 0,
              nrow = 8,
              ncol = 20) # create matrix with zeros
mat[
  sample(x = seq_len(length.out = 8*20),
         size = 15)
] <- seq(1, 15) # add non-zero entries
spmat <- as(object = mat, 
            Class = "TsparseMatrix") # make it sparse
path <- tempfile() # get temp file path

# The array will be 8x40 with dims "rows" and "cols" and domains [1,8] and [1,40]
dom <- tiledb::tiledb_domain(dims = c(tiledb::tiledb_dim(name = "rows",
                                                         domain = c(1L, 8L),
                                                         tile = 8L, 
                                                         type = "INT32"),
                                      tiledb::tiledb_dim(name = "cols", 
                                                         domain = c(1L, 40L),
                                                         tile = 40L, 
                                                         type = "INT32")))

# The array will be SPARSE with a single attribute "a" so
# each cell (i,j) cell can store an integer.
schema <- tiledb::tiledb_array_schema(domain = dom, 
                                      attrs = c(tiledb::tiledb_attr(name = "a",
                                                                    type = "INT32")),
                                      sparse = TRUE)
# Create the (empty) array on disk.
tiledb::tiledb_array_create(uri = path, 
                            schema = schema)

tiledb::fromDataFrame(obj = spmat, 
                      uri = path,
                      mode = "append")

tiledb::fromSparseMatrix(obj = spmat, 
                         uri = path,
                         mode = "append")
```

**Comment by eddelbuettel** [comment]
I may not have time right now to debug your example but here is a simpler (?) one each continuous integration test via unit tests executes:

```r
library(palmerpenguins)
data <- penguins

uri <- tempfile()

fromDataFrame(data, uri, col_index=1:2, mode="schema_only")
arr <- tiledb_array(uri, return_as="data.frame")
chk <- arr[]
expect_equal(nrow(chk), 0)              # no data
expect_equal(ncol(chk), ncol(data))     # but all columns

fromDataFrame(data, uri, col_index=1:2, mode="append")
arr <- tiledb_array(uri, return_as="data.frame")
chk <- arr[]
expect_equal(nrow(chk), nrow(data))     # all data
expect_equal(ncol(chk), ncol(data))     # all columns
```

I have others too if you want to explicitly write a schema.

**Comment by PedroMilanezAlmeida** [comment]
Thank you very much for your prompt help, in particular given how much else you do for the community.

Sorry that I was unclear in my previous post. The issue was that `fromDataFrame` cannot handle a sparse matrix, while `fromSparseMatrix` cannot handle `mode = "append"`.

So, I am now pre-processing the sparse matrices to the triplet format and storing each of them as a `data.frame`.

**Comment by eddelbuettel** [comment]
Ah, yes, I see. Spot on.  We do however have a less-used `fromSparseMatrix` which, unless I am rushing and overlooking something, could get a similar 'modal' treatment.  Would that help?  Do you want to take a stab at a PR?

**Comment by PedroMilanezAlmeida** [comment]
I admire your trust, but I am afraid I am far from technically capable of starting a PR myself.
